src/main/java/com/google/sps/servlets/RecipeServlet.java

```diff
@@ -14,22 +14,69 @@
 
 package com.google.sps.servlets;
 
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.FirebaseDatabase;
+import com.google.firebase.database.ValueEventListener;
 import com.google.gson.Gson;
+import com.google.sps.data.DatabaseReferences;
+import com.google.sps.data.Post;
+import java.io.BufferedReader;
 import java.io.IOException;
+import java.lang.StringBuilder;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.Collectors;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 @WebServlet("/api/post")
 public class RecipeServlet extends HttpServlet {
+  private Gson gson;
+
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {}
+  public void init() {
+    gson = new Gson();
+  }
 
   @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {}
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatabaseReferences.ROOT.child("post-metadata/0").addListenerForSingleValueEvent(new ValueEventListener() {
+      @Override
+      public void onDataChange(DataSnapshot dataSnapshot) {
```

Comment:
Umm, not sure I understood how this suppose to work. Looks like this is a listener for changes in the database, which is useful if you want to watch what's being added to the database in real time. Are you sure this is what you want? If you just want to get what's already in the database, firebase.database.Query seems to be what you want.

Reply:
You're right on, except their provided helper 'addListenerForSingleValueEvent' supposedly returns data as if we were issuing an asynchronous query, and only triggers the callback once. I'll look into firebase.database.Query, but I'm concerned it may be intended for use directly in JavaScript. If that's the case, I may just get rid of the servlet's doGet method entirely, and handle querying through the frontend.

Reply:
Query supports a bunch of different languages including Java: 
https://firebase.google.com/docs/firestore/query-data/queries#java_3

Reply:
I think this approach is intended for use with Firestore, I'm not sure if it works with Realtime DB. However, we may convert over to Firestore as we haven't invested much into Realtime DB at this point

Reply:
@hanpeiz that link is for firestore not realtimedb though?  although rtdb has a similar query interface, but still uses the callbacks.

I assume you're trying to do this with the "Admin" library, yeah? So https://firebase.google.com/docs/database/admin/retrieve-data is relevant. Which claims "blocking reads" in *some languages*. The key here is that Java isn't one of those languages, so you need to deal with listeners.

You're going to need to wrap the entire thing into a Future or similar (Promise is the same, but I've never used a Promise library on java serverside), and block until the future/promise completes, and then return that. it's going to be a bit of a pain. an alternative is, for data that you are just "passing through" without any manipulation on the server, to just query FB directly from the JS (which is sort of how they mostly intend you to do it, actually...)

But if you want to have the server merge stuff, or do transformations, or ... whatever else that the client can't easily do, then I suspect yes you'll need to wrap in a Future.

You should read this article: https://medium.com/google-cloud/firebase-asynchronous-operations-with-admin-java-sdk-82ca9b4f6022 -- not sure it really has the "answer" but looks like it covers some useful ground.

Seems like you may be able to do something like getValueAsync() which will return an ApiFuture that you can then .get() on (which will block until the result). Or something like that 👍

Reply:
Thank you both for the help and resources! I'm checking out the medium article and it looks very promising, but we may end up using Firestore actually, as it seems to have a richer set of features for performing queries when we have a set of search parameters. I'll update here with any decisions.

Reply:
I didn't have the courage to take on futures so the latest push uses firestore instead... We decided as a group we want to use this instead.